File: RTSim/RTRobot.m
classdef RTRobot < RTSim
    %RTRobot Controller for RTSim real-time robot simulation.
    %    https://github.com/FJFranklin/wifi-py-rpi-car-controller/tree/master/RTSim
    
    properties
        % list of global (class) variables used by setup(), loop() and ping_receive()
        target
        last_ping_time
        last_ping_distance
        position
        orientation
    end
    
    methods
        function obj = RTRobot (seconds, test_name)
            % This is the Matlab version of the coursework 'Matlab Robot':
            % In the following line, replace the number with your Student ID
            id_number = 170000000;

            if (nargin() < 1) % default to 180s (3 minutes)
                seconds = 180;
            end
            if (nargin() < 2) % default to simple layout
                test_name = 'default';
                % other options are: 'random', 'TNT', 'CWC' & 'BSB'
            end
            obj@RTSim (seconds, test_name, id_number);
        end
        function setup (obj) % setup() is called once at the beginning
            obj.target = obj.get_target ();       % where we are trying to get to

            % For example:
            obj.last_ping_time = 0;               % see ping_receive()
            obj.last_ping_distance = -1;
        end
        function loop (obj)
            % loop() is called repeatedly

            % For example:
            currentTime = obj.millis () / 1000;

            obj.position = obj.get_GPS ();        % roughly where we are
            obj.orientation = obj.get_compass (); % which direction we are looking

            if (currentTime > 4)
                obj.ping_send ();                 % it will not actually send more often than every 0.1s
            end

            obj.set_ping_angle (180);
            obj.set_wheel_speeds (-127, -126);
        end
        function ping_receive (obj, distance)
            % response to an obj.ping_send ()

            % For example:
            obj.last_ping_time = obj.millis ();   % the last time we received a ping [in milliseconds]
            obj.last_ping_distance = distance;    % distance measured (-ve if no echo)

            if (distance >= 0)                    % a -ve distance implies nothing seen
                disp (['position=(',num2str(obj.position(1)),',',num2str(obj.position(2)),...
                       '), orientation=',num2str(obj.orientation),...
                       '; distance=',num2str(distance)]);
            end
        end
    end
    
end

File: RTSim/RTRobot.py
from RTSim import RTSim

class RTRobot(RTSim):
    """
    RTRobot Controller for RTSim real-time robot simulation.
    https://github.com/FJFranklin/wifi-py-rpi-car-controller/tree/master/RTSim
    """

    def __init__(self, seconds=180, test_name='default'):
        # usage: RTRobot (seconds, test_name)
        # where test_name is one of 'default', 'random', 'TNT', 'CWC' or 'BSB'

        # This is the Python version of the coursework 'Matlab Robot':
        # In the following line, replace the number with your Student ID
        id_number = 170000000;

        RTSim.__init__(self, seconds, test_name, id_number)

    def setup(self):
        # setup() is called once at the beginning

        self.target = self.get_target()       # where we're trying to get to

        # For example:
        self.last_ping_time = 0               # see ping_receive()
        self.last_ping_distance = -1

        # To work out which of the trials we're running:
        results_so_far = self.get_result()
        test_name = results_so_far['Trial']
        print('This trial is:', test_name)

    def loop(self):
        # loop() is called repeatedly

        # For example:
        currentTime = self.millis() / 1000

        self.position = self.get_GPS()        # roughly where we are
        self.orientation = self.get_compass() # which direction we're looking

        if currentTime > 4:
            self.ping_send()                  # it won't actually send more often than every 0.1s

        self.set_ping_angle(180)
        self.set_wheel_speeds(-127, -126)

    def ping_receive(self, distance):
        # response to an self.ping_send()

        # For example:
        self.last_ping_time = self.millis()  # the last time we received a ping [in milliseconds]
        self.last_ping_distance = distance   # distance measured (-ve if no echo)

        if distance >= 0:                    # a -ve distance implies nothing seen
            print('position=(', self.position[0], ',', self.position[1], '), orientation=', self.orientation, '; distance=', distance, sep='')

if __name__ == "__main__":
    # Option to run from command line

    import argparse

    parser = argparse.ArgumentParser(description="RTRobot Coursework - Guide a two-wheeled robot round the map.")

    parser.add_argument('--duration', help='How many seconds to run [40].', default=40, type=int)
    parser.add_argument('--trial',    help='Specify map type [default].',   default='default', choices=['default', 'random', 'TNT', 'CWC', 'BSB'])

    args = parser.parse_args()

    R = RTRobot(args.duration, args.trial)
    print(R.get_result())
